File: openharmony-sig/ohos_react_native/docs/Samples/Sample/SampleApp/entry/src/main/ets/TurboModule/SampleTurboModulePackage.ts
/**
 * Copyright (c) 2024 Huawei Technologies Co., Ltd.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree.
 */

import { RNPackage, TurboModulesFactory } from '@rnoh/react-native-openharmony/ts';
import { TurboModule, TurboModuleContext } from '@rnoh/react-native-openharmony/ts';
import { SampleTurboModule } from './SampleTurboModule';
import { SampleTurboModule2 } from './SampleTurboModule2';

const SAMPLE_TURBO_MODULE_CLASS_BY_NAME: Record<string, typeof TurboModule> = {
  [SampleTurboModule.NAME]: SampleTurboModule,
  [SampleTurboModule2.NAME]: SampleTurboModule2,
}

class SampleTurboModulesFactory extends TurboModulesFactory {
  createTurboModule(name: string): TurboModule | null {
    if (this.hasTurboModule(name)) {
      return new SAMPLE_TURBO_MODULE_CLASS_BY_NAME[name](this.ctx);
    }
    return null;
  }

  hasTurboModule(name: string): boolean {
    return name in SAMPLE_TURBO_MODULE_CLASS_BY_NAME;
  }
}

export class SampleTurboModulePackage extends RNPackage {
  createTurboModulesFactory(ctx: TurboModuleContext): TurboModulesFactory {
    return new SampleTurboModulesFactory(ctx);
  }
}

File: openharmony-sig/ohos_react_native/docs/Samples/using_turboModule/SampleApp/entry/src/main/cpp/SampleTurboModulePackage.cpp
/**
 * Copyright (c) 2024 Huawei Technologies Co., Ltd.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree.
 */

#include "SampleTurboModulePackage.h"
#include "./TurboModules/SampleTurboModuleSpec.h"
#include "./TurboModules/SampleTurboModule2Spec.h"
#include "./TurboModules/SampleWorkerTurboModuleSpec.h"

using namespace rnoh;
using namespace facebook;

class SampleTurboModuleFactoryDelegate : public TurboModuleFactoryDelegate {
    
    public:
    SharedTurboModule createTurboModule(Context ctx, const std::string &name) const override {
        if (name == "SampleTurboModule") {
            return std::make_shared<NativeSampleTurboModuleSpecJSI>(ctx, name);
        } else if (name == "SampleTurboModule2") {
            return std::make_shared<NativeSampleTurboModule2SpecJSI>(ctx, name);    
        }else if (name == "SampleWorkerTurboModule"){
            return std::make_shared<NativeSampleWorkerTurboModuleSpecJSI>(ctx, name);
        }
        return nullptr;
    };
};

class ButtonViewPackageComponentInstanceFactoryDelegate : public ComponentInstanceFactoryDelegate {
    public:
        using ComponentInstanceFactoryDelegate::ComponentInstanceFactoryDelegate;
    
        ComponentInstance::Shared create(ComponentInstance::Context ctx) override {
            return nullptr;
        }
};

std::unique_ptr<TurboModuleFactoryDelegate> SampleTurboModulePackage::createTurboModuleFactoryDelegate() {
    return std::make_unique<SampleTurboModuleFactoryDelegate>();
}
